File: java/src/main/java/euler/java/Problem3.java
package euler.java;

/**
 * Problem 3 (the Largest Prime Factor)
 * <p/>
 * The prime factors of 13195 are 5, 7, 13 and 29.<br />
 * What is the largest prime factor of the number 600851475143 ?
 *
 * @author Ryo TANAKA
 * @since 1.0
 */
public class Problem3 {
    public long getLargestPrimeFactor(long num) {
        // とりあえず、longの範囲では100あれば十分
        long[] primes = new long[100];
        int idx = 0;
        long div = num;
        for (long i = 2; i < div; i++) {
            if (num % i == 0) {
                primes[idx] = i;
                idx++;
                div = num / i;
                primes[idx] = div;
                idx++;
            }
        }
        long max = 0L;
        for (int i = 0; i < primes.length - 1 && primes[i] != 0; i++) {
            long candidate = primes[i];
            if (max < candidate && !isDividable(candidate, primes)) {
                max = candidate;
            }
        }
        return max != 0 ? max : num;
    }

    private boolean isDividable(long l, long[] primes) {
        for (int i = 0; i < primes.length - 1 && primes[i] != 0; i++) {
            long prime = primes[i];
            if (l != prime && l % prime == 0) {
                return true;
            }
        }
        return false;
    }
}

File: scala/src/main/scala/euler/scala/Problem3.scala
package euler.scala

/**
 * Problem 3 (the Largest Prime Factor)
 * <p/>
 * The prime factors of 13195 are 5, 7, 13 and 29.<br />
 * What is the largest prime factor of the number 600851475143 ?
 *
 * @author Ryo TANAKA
 * @since 1.0
 */
object Problem3 {
  def maxPrime(num: Long) = {
    var limit = num
    lazy val primes: Stream[Long] =
      2 #:: primes.map(x => {
        // Stream#range(start, end) excludes end value.
        Stream.range(x, limit + 1).find(limit % _ == 0) match {
          case Some(prime) => limit = limit / prime; prime
          case None => x
        }
      })
    primes.take(10).last
  }
}
